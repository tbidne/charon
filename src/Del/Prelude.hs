-- | Custom prelude.
--
-- @since 0.1
module Del.Prelude
  ( module X,
  )
where

import Control.Applicative as X
  ( Alternative ((<|>)),
    Applicative (pure, (<*>)),
    (*>),
  )
import Control.DeepSeq as X (NFData)
import Control.Exception as X
  ( Exception (displayException, fromException, toException),
    SomeAsyncException (SomeAsyncException),
    SomeException,
    bracket,
    catch,
    finally,
    throwIO,
  )
import Control.Monad as X (Monad ((>>=)), join, (<=<), (>=>))
import Control.Monad.Fail as X (MonadFail (fail))
import Data.Bool as X (Bool (False, True), not, otherwise, (&&))
import Data.ByteString as X (ByteString)
import Data.Either as X (Either (Left, Right))
import Data.Eq as X (Eq ((/=), (==)))
import Data.Foldable as X (Foldable (foldl', foldr, length), for_)
import Data.Function as X (const, flip, ($), (.))
import Data.Functor as X (Functor (fmap), ($>), (<$>))
import Data.HashMap.Strict as X (HashMap)
import Data.HashSet as X (HashSet)
import Data.Hashable as X (Hashable (hashWithSalt))
import Data.IORef as X (IORef, modifyIORef', newIORef, readIORef)
import Data.Int as X (Int)
import Data.Kind as X (Type)
import Data.List as X (zipWith)
import Data.List.NonEmpty as X (NonEmpty ((:|)))
import Data.Maybe as X (Maybe (Just, Nothing), maybe)
import Data.Monoid as X (Monoid (mconcat, mempty))
import Data.Ord as X (Ord (compare, (<=), (>), (>=)), Ordering (EQ, GT, LT))
import Data.Proxy as X (Proxy (Proxy))
import Data.Semigroup as X (Semigroup ((<>)))
import Data.String as X (IsString (fromString), String)
import Data.Text as X (Text)
import Data.Traversable as X (traverse)
import Data.Tuple as X (curry, uncurry)
import Data.Vector as X (Vector)
import Data.Word as X (Word16)
import GHC.Enum as X (Bounded (maxBound, minBound))
import GHC.Err as X (error, undefined)
import GHC.Float as X (Double)
import GHC.Generics as X (Generic)
import GHC.Integer as X (Integer)
import GHC.Natural as X (Natural)
import GHC.Num as X (Num ((+), (-)))
import GHC.Real as X (fromIntegral)
import GHC.Stack as X (HasCallStack)
import Optics.Core as X (Iso', iso, review, view, (%), (^.), _1, _2)
import Prettyprinter as X
  ( Pretty (pretty),
    layoutCompact,
    line,
    vsep,
    (<+>),
  )
import Prettyprinter.Render.Text as X (renderStrict)
import System.Directory as X (createDirectoryIfMissing)
import System.Exit as X (exitFailure)
import System.FilePath as X ((</>))
import System.IO as X
  ( BufferMode (NoBuffering),
    FilePath,
    IO,
    print,
    putStr,
    putStrLn,
  )
import Text.Show as X (Show (show))
