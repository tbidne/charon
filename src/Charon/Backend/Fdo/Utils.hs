module Charon.Backend.Fdo.Utils
  ( percentEncodeFileName,
  )
where

import Charon.Data.Paths (PathI (MkPathI), PathIndex (TrashEntryFileName))
import Charon.Prelude
import Charon.Utils qualified as Utils
import Data.Text qualified as T

percentEncodeFileName ::
  forall m pd k.
  ( HasCallStack,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    MonadThrow m
  ) =>
  pd ->
  m ByteString
percentEncodeFileName pd =
  percentEncode <$> decodeOsToFpThrowM fileName
  where
    percentEncode = Utils.percentEncode . encodeUtf8 . T.pack
    MkPathI fileName = view #fileName pd
