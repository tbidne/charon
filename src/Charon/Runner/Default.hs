module Charon.Runner.Default
  ( Default (..),
    fromDefault,
    (<|.|>),
  )
where

import Charon.Prelude

-- | For types with a default value.
class Default a where
  def :: a

instance Default (Maybe a) where
  def = Nothing

fromDefault :: (Default a) => Maybe a -> a
fromDefault = fromMaybe def

(<|.|>) :: (Default a) => Maybe a -> Maybe a -> a
x <|.|> y = fromDefault (x <|> y)

infixr 6 <|.|>
