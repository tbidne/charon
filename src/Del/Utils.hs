-- | Provides internal utility functions
--
-- @since 0.1
module Del.Utils
  ( allM1,
  )
where

import Del.Prelude

-- | 'allM' that must have at least one 'True'.
--
-- @since 0.1
allM1 :: Monad m => NonEmpty (m Bool) -> m Bool
allM1 (m :| ms) =
  m >>= \case
    True -> allM ms
    False -> pure False

-- | 'all' lifted to monads.
--
-- @since 0.1
allM :: (Foldable t, Monad m) => t (m Bool) -> m Bool
allM = foldr f (pure True)
  where
    f m acc =
      m >>= \case
        True -> acc
        False -> pure False
