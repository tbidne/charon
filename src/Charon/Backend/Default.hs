{-# LANGUAGE TemplateHaskell #-}

-- | This module (and sub-hierarchy) provides logic that can be used by any
-- backend (i.e. cbor, fdo, json) that has the following structure:
--
-- @
--   trash/paths/ (paths)
--   trash/info/  (info files)
-- @
--
-- The only difference is in the (de)serialization.
module Charon.Backend.Default
  ( -- * Delete
    delete,
    deletePostHook,
    permDelete,
    permDeletePostHook,
    emptyTrash,

    -- * Restore
    restore,
    restorePostHook,

    -- * Information
    getIndex,
    getMetadata,

    -- * Transformations
    merge,
  )
where

import Charon.Backend.Default.BackendArgs (BackendArgs)
import Charon.Backend.Default.Index qualified as Default.Index
import Charon.Backend.Default.Trash qualified as Trash
import Charon.Backend.Default.Utils qualified as Default.Utils
import Charon.Class.Serial (Serial (DecodeExtra))
import Charon.Data.Index (Index)
import Charon.Data.Index qualified as Index
import Charon.Data.Metadata (Metadata (MkMetadata))
import Charon.Data.Metadata qualified as Metadata
import Charon.Data.PathData (PathData)
import Charon.Data.PathType (PathTypeW)
import Charon.Data.Paths
  ( PathI (MkPathI),
    PathIndex
      ( TrashEntryFileName,
        TrashEntryOriginalPath,
        TrashEntryPath,
        TrashHome
      ),
  )
import Charon.Data.Paths qualified as Paths
import Charon.Data.Timestamp (Timestamp (MkTimestamp))
import Charon.Data.UniqueSeqNE (UniqueSeqNE)
import Charon.Data.UniqueSeqNE qualified as USeqNE
import Charon.Env (HasTrashHome (getTrashHome))
import Charon.Env qualified as Env
import Charon.Prelude
import Charon.Utils qualified as Utils
import Data.Bytes qualified as Bytes
import Data.Char qualified as Ch
import Data.Text qualified as T
import Effects.System.Terminal qualified as Term
import Effects.Time (getSystemTime)
import Numeric.Algebra.Additive.AMonoid (AMonoid (zero))
import Numeric.Algebra.Additive.ASemigroup (ASemigroup ((.+.)))
import Numeric.Literal.Rational (FromRational (afromRational))

-- NOTE: For functions that can encounter multiple exceptions, the first
-- one is rethrown.

-- | @delete trash p@ moves path @p@ to the given trash location @trash@ and
-- writes an entry in the trash index. If the trash location is not given,
-- defaults to XDG data e.g. @~\/.local/share/charon/@.
delete ::
  forall m env pd k.
  ( HasCallStack,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    LabelOptic' "originalPath" k pd (PathI TrashEntryOriginalPath),
    HasTrashHome env,
    MonadCatch m,
    MonadFileWriter m,
    MonadIORef m,
    MonadLoggerNS m,
    MonadPathWriter m,
    MonadReader env m,
    MonadTerminal m,
    MonadTime m,
    Serial pd,
    Show pd
  ) =>
  BackendArgs m pd ->
  UniqueSeqNE (PathI TrashEntryOriginalPath) ->
  m ()
delete backendArgs = deletePostHook backendArgs (const $ pure ())

-- | 'delete' that takes a callback that runs on the created path data, assuming
-- the delete succeeded.
deletePostHook ::
  forall m env pd k.
  ( HasCallStack,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    LabelOptic' "originalPath" k pd (PathI TrashEntryOriginalPath),
    HasTrashHome env,
    MonadCatch m,
    MonadFileWriter m,
    MonadIORef m,
    MonadLoggerNS m,
    MonadPathWriter m,
    MonadReader env m,
    MonadTerminal m,
    MonadTime m,
    Serial pd,
    Show pd
  ) =>
  BackendArgs m pd ->
  ((pd, PathTypeW, PathI TrashEntryPath) -> m ()) ->
  UniqueSeqNE (PathI TrashEntryOriginalPath) ->
  m ()
deletePostHook backendArgs postHook paths = addNamespace "deletePostHook" $ do
  $(logDebug) $ "Paths: " <> USeqNE.display Paths.toText paths
  trashHome <- asks getTrashHome

  void Trash.createTrash

  someExRef <- newIORef Nothing
  currTime <- MkTimestamp <$> getSystemTime

  let deleteAction = Trash.mvOriginalToTrash backendArgs trashHome currTime

      handleEx p ex = do
        $(logError) (T.pack $ displayNoCS ex)
        putStrLn
          $ mconcat
            [ "Error deleting path '",
              decodeOsToFpDisplayEx $ p ^. #unPathI,
              "': ",
              displayNoCS ex
            ]
        writeIORef someExRef (Just ex)

  -- move paths to trash
  for_ paths $ \p -> do
    tryAnyCS (deleteAction p) >>= \case
      Left ex -> handleEx p ex
      Right pd -> postHook pd

  Utils.throwIfEx someExRef

-- | Permanently deletes the paths from the trash.
permDelete ::
  forall m env pd k.
  ( DecodeExtra pd ~ PathI TrashEntryFileName,
    HasCallStack,
    HasTrashHome env,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    MonadAsync m,
    MonadCatch m,
    MonadFileReader m,
    MonadHandleWriter m,
    MonadIORef m,
    MonadPathReader m,
    MonadPathWriter m,
    MonadLoggerNS m,
    MonadReader env m,
    MonadTerminal m,
    Serial pd,
    Show pd
  ) =>
  BackendArgs m pd ->
  Bool ->
  UniqueSeqNE (PathI TrashEntryFileName) ->
  m ()
permDelete backendArgs = permDeletePostHook backendArgs (const $ pure ())

-- | Permanently deletes the paths from the trash.
permDeletePostHook ::
  forall m env pd k.
  ( DecodeExtra pd ~ PathI TrashEntryFileName,
    HasCallStack,
    HasTrashHome env,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    MonadAsync m,
    MonadCatch m,
    MonadFileReader m,
    MonadHandleWriter m,
    MonadIORef m,
    MonadPathReader m,
    MonadPathWriter m,
    MonadLoggerNS m,
    MonadReader env m,
    MonadTerminal m,
    Serial pd,
    Show pd
  ) =>
  BackendArgs m pd ->
  (PathData -> m ()) ->
  Bool ->
  UniqueSeqNE (PathI TrashEntryFileName) ->
  m ()
permDeletePostHook backendArgs postHook force paths = addNamespace "permDeletePostHook" $ do
  $(logDebug) $ "Paths: " <> USeqNE.display Paths.toText paths
  trashHome <- asks getTrashHome

  someExRef <- newIORef Nothing

  -- permanently delete paths
  addNamespace "deleting" $ for_ paths $ \p ->
    -- Record error if any occurred
    (Trash.permDeleteFromTrash backendArgs postHook force trashHome p >>= Utils.setRefIfJust someExRef)
      `catchAnyCS` \ex -> do
        $(logError) (T.pack $ displayNoCS ex)
        putStrLn
          $ mconcat
            [ "Error permanently deleting path '",
              decodeOsToFpDisplayEx $ p ^. #unPathI,
              "': ",
              displayNoCS ex
            ]
        -- in case Trash.permDeleteFromTrash throws an exception
        writeIORef someExRef (Just ex)

  Utils.throwIfEx someExRef

-- | Reads the index at either the specified or default location. If the
-- file does not exist, returns empty.
getIndex ::
  forall m env pd k.
  ( DecodeExtra pd ~ PathI TrashEntryFileName,
    HasCallStack,
    HasTrashHome env,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    MonadCatch m,
    MonadFileReader m,
    MonadPathReader m,
    MonadLoggerNS m,
    MonadReader env m,
    Serial pd
  ) =>
  BackendArgs m pd ->
  m Index
getIndex backendArgs = addNamespace "getIndex" $ do
  Trash.doesTrashExist >>= \case
    True -> Default.Index.readIndex backendArgs
    False -> do
      $(logTrace) "Trash does not exist."
      pure Index.empty

-- | Retrieves metadata for the trash directory.
getMetadata ::
  forall m env pd k.
  ( DecodeExtra pd ~ PathI TrashEntryFileName,
    HasCallStack,
    HasTrashHome env,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    MonadCatch m,
    MonadFileReader m,
    MonadLoggerNS m,
    MonadPathReader m,
    MonadReader env m,
    Serial pd
  ) =>
  BackendArgs m pd ->
  m Metadata
getMetadata backendArgs = addNamespace "getMetadata" $ do
  $(logTrace) "In getMetadata"
  trashHome <- asks getTrashHome
  trashLog <- Env.getTrashLog

  let MkPathI trashPathsDir = Default.Utils.getTrashPathDir trashHome

  Trash.doesTrashExist >>= \case
    False -> do
      $(logTrace) "Trash does not exist."
      pure Metadata.empty
    True -> do
      -- Index size
      index <- view #unIndex <$> Default.Index.readIndexTrashHome backendArgs trashHome
      let numIndex = length index
      $(logDebug) ("Index size: " <> showt numIndex)

      -- Num entries
      numEntries <- foldl' (\acc _ -> acc + 1) 0 <$> listDirectory trashPathsDir
      $(logDebug) ("Num entries: " <> showt numEntries)

      -- Log size
      let logPath = trashLog ^. #unPathI
      logExists <- doesFileExist logPath
      logSize <-
        if logExists
          then Bytes.normalize . toDouble . MkBytes @B <$> getFileSize logPath
          else do
            $(logTrace) "Log does not exist"
            pure (afromRational 0)

      -- Summed size
      allFiles <- Utils.getAllFiles trashPathsDir
      let allSizes = foldl' (\acc (pd, _) -> (pd ^. #size) .+. acc) zero index
          numFiles = length allFiles
          size = fromIntegral <$> Bytes.normalize allSizes

      $(logDebug) ("Num all files: " <> showt numFiles)
      $(logDebug) ("Total size: " <> showt size)

      -- NOTE: If the index is successfully read then we have verified that
      -- all invariants are preserved i.e. bijection between /files and /info.

      pure
        $ MkMetadata
          { numEntries = toNat numEntries,
            numFiles = toNat numFiles,
            logSize,
            size
          }
  where
    toDouble :: (Integral a) => Bytes s a -> Bytes s Double
    toDouble = fmap fromIntegral
    toNat :: Int -> Natural
    toNat = fromIntegral

-- | @restore trash p@ restores the trashed path @\<trash\>\/p@ to its original
-- location.
restore ::
  forall m env pd k.
  ( DecodeExtra pd ~ PathI TrashEntryFileName,
    HasCallStack,
    HasTrashHome env,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    MonadCatch m,
    MonadFileReader m,
    MonadIORef m,
    MonadLoggerNS m,
    MonadPathReader m,
    MonadPathWriter m,
    MonadReader env m,
    MonadTerminal m,
    Serial pd,
    Show pd
  ) =>
  BackendArgs m pd ->
  UniqueSeqNE (PathI TrashEntryFileName) ->
  m ()
restore backendArgs = restorePostHook backendArgs (const $ pure ())

-- | @restore trash p@ restores the trashed path @\<trash\>\/p@ to its original
-- location.
restorePostHook ::
  forall m env pd k.
  ( DecodeExtra pd ~ PathI TrashEntryFileName,
    HasCallStack,
    HasTrashHome env,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    MonadCatch m,
    MonadFileReader m,
    MonadIORef m,
    MonadLoggerNS m,
    MonadPathReader m,
    MonadPathWriter m,
    MonadReader env m,
    MonadTerminal m,
    Serial pd,
    Show pd
  ) =>
  BackendArgs m pd ->
  (PathData -> m ()) ->
  UniqueSeqNE (PathI TrashEntryFileName) ->
  m ()
restorePostHook backendArgs postHook paths = addNamespace "restorePostHook" $ do
  $(logDebug) $ "Paths: " <> USeqNE.display Paths.toText paths
  trashHome <- asks getTrashHome

  someExRef <- newIORef Nothing

  -- move trash paths back to original location
  addNamespace "restoring" $ for_ paths $ \p ->
    -- Record error if any occurred
    (Trash.restoreTrashToOriginal backendArgs postHook trashHome p >>= Utils.setRefIfJust someExRef)
      `catchAnyCS` \ex -> do
        $(logError) (T.pack $ displayNoCS ex)
        putStrLn
          $ mconcat
            [ "Error restoring path '",
              decodeOsToFpDisplayEx $ p ^. #unPathI,
              "': ",
              displayNoCS ex
            ]
        writeIORef someExRef (Just ex)

  Utils.throwIfEx someExRef

-- | Empties the trash.
emptyTrash ::
  forall m env pd k.
  ( DecodeExtra pd ~ PathI TrashEntryFileName,
    HasCallStack,
    HasTrashHome env,
    Is k A_Getter,
    LabelOptic' "fileName" k pd (PathI TrashEntryFileName),
    MonadCatch m,
    MonadFileReader m,
    MonadHandleWriter m,
    MonadLoggerNS m,
    MonadPathReader m,
    MonadPathWriter m,
    MonadReader env m,
    MonadTerminal m,
    Serial pd
  ) =>
  BackendArgs m pd ->
  Bool ->
  m ()
emptyTrash backendArgs force = addNamespace "emptyTrash" $ do
  $(logTrace) "In emptyTrash"
  trashHome@(MkPathI th) <- asks getTrashHome

  exists <- doesDirectoryExist th
  if not exists
    then do
      $(logTrace) "Trash home does not exist."
      putTextLn $ Paths.toText trashHome <> " is empty."
    else
      if force
        then do
          $(logTrace) "Force on; deleting entire trash."
          removeDirectoryRecursive th
          void Trash.createTrash
        else do
          Utils.noBuffering
          metadata <- getMetadata backendArgs
          putStrLn ""
          putTextLn $ Utils.renderPretty metadata
          putStr "Permanently delete all contents (y/n)? "
          c <- Ch.toLower <$> Term.getChar
          if
            | c == 'y' -> do
                $(logTrace) "Deleting contents."
                removeDirectoryRecursive th
                void Trash.createTrash
                putStrLn ""
            | c == 'n' -> do
                $(logTrace) "Not deleting contents."
                putStrLn ""
            | otherwise -> putStrLn ("\nUnrecognized: " <> [c])

merge ::
  ( HasCallStack,
    MonadFileReader m,
    MonadIORef m,
    MonadLoggerNS m,
    MonadMask m,
    MonadPathReader m,
    MonadPathWriter m
  ) =>
  PathI TrashHome ->
  PathI TrashHome ->
  m ()
merge src dest = addNamespace "merge" $ Trash.mergeTrashDirs src dest
