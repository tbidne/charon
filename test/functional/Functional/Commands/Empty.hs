{-# LANGUAGE QuasiQuotes #-}

-- | Tests for m command.
module Functional.Commands.Empty
  ( tests,
  )
where

import Charon.Backend.Default.Utils qualified as Default.Utils
import Functional.Prelude

-- NOTE: These tests currently rely on internal details for the trash
-- structure (see the usage of Default.Utils). If we ever get a non-compliant
-- backend, this will have to change.

tests :: IO TestEnv -> TestTree
tests testEnv =
  testGroup
    "Empty Command"
    [ emptyTrash testEnv',
      emptyTrashTwice testEnv',
      emptyPrompt testEnv',
      missingInfoForcesDelete testEnv',
      missingPathsForcesDelete testEnv'
    ]
  where
    testEnv' = appendTestDir "empty" <$> testEnv

emptyTrash :: IO TestEnv -> TestTree
emptyTrash getTestEnv =
  testGoldenParams
    $ MkGoldenParams
      { runner,
        testDesc = "Empties trash",
        testName = testDirPrefix <> [osp|emptyTrash|]
      }
  where
    runner = do
      testEnv <- getTestEnv
      usingReaderT testEnv $ appendTestDirM "emptyTrash" $ do
        testDir <- getTestDir

        let filesToDelete = (testDir </>!) <$> ["f1", "f2", "f3"]
            dirsToDelete = (testDir </>!) <$> ["dir1", "dir2", "dir4"]
            fileLinkToDelete = testDir </> [osp|file-link|]
            dirLinkToDelete = testDir </> [osp|dir-link|]
            linksToDelete = [fileLinkToDelete, dirLinkToDelete]

        delArgList <- withSrArgsPathsM ["delete"] (filesToDelete <> dirsToDelete <> linksToDelete)

        -- setup
        -- test w/ a nested dir
        createDirectories (dirsToDelete <> [testDir </>! "dir2/dir3", testDir </>! "dir4"])
        -- test w/ a file in dir
        createFiles ((testDir </>! "dir2/dir3/foo") : filesToDelete)
        createSymlinks [F fileLinkToDelete, D dirLinkToDelete, F $ testDir </>! "dir4" </>! "link"]
        assertPathsExist (filesToDelete ++ dirsToDelete)
        assertSymlinksExist linksToDelete

        runCharon delArgList

        -- file assertions
        assertPathsDoNotExist (filesToDelete ++ dirsToDelete)

        -- trash structure assertions
        bs1 <- captureIndexBs testDir
        assertFdoDirectorySizesM ["dir1", "dir2", "dir4"]

        -- EMPTY

        emptyArgList <- withSrArgsM ["empty", "--prompt", "off"]
        runCharon emptyArgList

        -- file assertions
        assertPathsDoNotExist (filesToDelete ++ dirsToDelete)

        -- trash structure assertions
        bs2 <- captureIndexBs testDir
        pure $ bs1 <> bs2

emptyTrashTwice :: IO TestEnv -> TestTree
emptyTrashTwice getTestEnv = testCase "Calling empty twice does not error" $ do
  testEnv <- getTestEnv
  usingReaderT testEnv $ appendTestDirM "emptyTrashTwice" $ do
    emptyArgs <- withSrArgsM ["empty", "--prompt", "off"]
    runCharon emptyArgs
    runCharon emptyArgs

emptyPrompt :: IO TestEnv -> TestTree
emptyPrompt getTestEnv =
  testGoldenParams
    $ MkGoldenParams
      { runner,
        testDesc = "Empties w/ no response deletes nothing",
        testName = testDirPrefix <> [osp|emptyPrompt|]
      }
  where
    runner = do
      testEnv <- getTestEnv
      usingReaderT testEnv $ appendTestDirM "emptyPrompt" $ do
        testDir <- getTestDir
        let fileDeleteNames = show @Int <$> [1 .. 5]
            fileDeletePaths = (testDir </>!) <$> fileDeleteNames

        delArgList <- withSrArgsPathsM ["delete"] fileDeletePaths

        -- setup
        -- test w/ a file in dir
        createFiles fileDeletePaths
        assertPathsExist fileDeletePaths

        runCharon delArgList

        -- file assertions
        assertPathsDoNotExist fileDeletePaths

        -- trash structure assertions
        bs1 <- captureIndexBs testDir

        -- EMPTY

        emptyArgList <- withSrArgsM ["empty"]
        runCharon emptyArgList

        -- trash structure assertions
        bs2 <- captureIndexBs testDir
        pure $ bs1 <> bs2

missingInfoForcesDelete :: IO TestEnv -> TestTree
missingInfoForcesDelete getTestEnv =
  testGoldenParams
    $ MkGoldenParams
      { runner,
        testDesc = "empty --prompt off overwrites bad directory (no info.)",
        testName = testDirPrefix <> [osp|missingInfoForcesDelete|]
      }
  where
    runner = do
      testEnv <- getTestEnv
      usingReaderT testEnv $ appendTestDirM "missingInfoForcesDelete" $ do
        testDir <- getTestDir

        let trashDir = testDir </> pathDotTrash
            filesToDelete = (testDir </>!) <$> ["f1", "f2", "f3"]
            dirsToDelete = (testDir </>!) <$> ["dir1", "dir2"]
        delArgList <- withSrArgsPathsM ["delete"] (filesToDelete <> dirsToDelete)

        -- setup
        -- test w/ a nested dir
        createDirectories (dirsToDelete <> [testDir </>! "dir2/dir3"])
        -- test w/ a file in dir
        createFiles ((testDir </>! "dir2/dir3/foo") : filesToDelete)
        assertPathsExist (filesToDelete ++ dirsToDelete)

        -- delete files
        runCharon delArgList

        -- file assertions
        assertPathsDoNotExist (filesToDelete ++ dirsToDelete)

        -- trash structure assertions
        bs1 <- captureIndexBs testDir

        -- delete info dir, leaving trash dir in bad state
        clearDirectory (trashDir </> Default.Utils.pathInfo)

        emptyArgList <- withSrArgsM ["empty", "--prompt", "off"]
        runCharon emptyArgList

        assertPathsExist $ fmap (trashDir </>) [Default.Utils.pathInfo, Default.Utils.pathFiles]

        -- trash structure assertions
        bs2 <- captureIndexBs testDir
        pure $ bs1 <> bs2

missingPathsForcesDelete :: IO TestEnv -> TestTree
missingPathsForcesDelete getTestEnv =
  testGoldenParams
    $ MkGoldenParams
      { runner,
        testDesc = "empty --prompt off overwrites bad directory (no paths/)",
        testName = testDirPrefix <> [osp|missingPathsForcesDelete|]
      }
  where
    runner = do
      testEnv <- getTestEnv
      usingReaderT testEnv $ appendTestDirM "missingPathsForcesDelete" $ do
        testDir <- getTestDir

        let trashDir = testDir </> pathDotTrash
            filesToDelete = (testDir </>!) <$> ["f1", "f2", "f3"]
            dirsToDelete = (testDir </>!) <$> ["dir1", "dir2"]

        delArgList <- withSrArgsPathsM ["delete"] (filesToDelete <> dirsToDelete)

        -- setup
        -- test w/ a nested dir
        createDirectories (dirsToDelete <> [testDir </>! "dir2/dir3"])
        -- test w/ a file in dir
        createFiles ((testDir </>! "dir2/dir3/foo") : filesToDelete)
        assertPathsExist (filesToDelete ++ dirsToDelete)

        -- delete files
        runCharon delArgList

        -- file assertions
        assertPathsDoNotExist (filesToDelete ++ dirsToDelete)

        -- trash structure assertions
        bs1 <- captureIndexBs testDir

        -- delete info dir, leaving trash dir in bad state
        clearDirectory (trashDir </> Default.Utils.pathFiles)

        emptyArgList <- withSrArgsM ["empty", "--prompt", "off"]
        runCharon emptyArgList

        -- file assertions
        assertPathsDoNotExist (filesToDelete ++ dirsToDelete)
        assertPathsExist $ fmap (trashDir </>) [Default.Utils.pathInfo, Default.Utils.pathFiles]

        -- trash structure assertions
        bs2 <- captureIndexBs testDir
        pure $ bs1 <> bs2

testDirPrefix :: OsString
testDirPrefix = [osstr|empty_|]
