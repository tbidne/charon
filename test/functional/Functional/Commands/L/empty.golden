TERMINAL LIST


Entries:      0
Total Files:  0
Log size:     0.00B
Size:         0.00B 



LOGS LIST
[2020-05-31 12:00:00][functional.getIndex][Debug][src/SafeRm.hs] Trash home: <dir>/safe-rm/functional/l1/.trash
[2020-05-31 12:00:00][functional.getIndex][Debug][src/SafeRm.hs] Trash does not exist.
[2020-05-31 12:00:00][functional.getMetadata][Debug][src/SafeRm.hs] Trash home: <dir>/safe-rm/functional/l1/.trash
[2020-05-31 12:00:00][functional.getMetadata][Info][src/SafeRm.hs] Trash does not exist.
